src/translate/pddl_parser/parsing_functions.py

```diff
@@ -188,121 +314,229 @@ def add_effect(tmp_effect, result):
+
+
+def parse_action(context, no, alist, type_dict, predicate_dict):
+    syntax = "(:action NAME " \
+             "[:parameters PARAMETERS]? " \
+             "[:precondition PRECONDITION]? " \
+             ":effect EFFECT)"
+    with context.layer(f"Parsing {no}. action"):
+        context.check_list(alist, min_size=4, finalmsg=f"Expecting '{syntax}'.")
+        iterator = iter(alist)
+        action_tag = next(iterator)
+        assert action_tag == ":action"
+        name = next(iterator)
+        context.check_string(name, item_name="Action name")
+    with context.layer(f"Parsing {no}. action '{name}'"):
         try:
-            cost = parse_effects(
-                effect_list, eff, type_dict, predicate_dict)
-        except ValueError as e:
-            raise SystemExit("Error in Action %s\nReason: %s." % (name, e))
-    for rest in iterator:
-        assert False, rest
+            with context.layer("Parsing parameters"):
+                parameters_tag_opt = next(iterator)
+                if parameters_tag_opt == ":parameters":
+                    parameters_list = next(iterator)
+                    context.check_list(parameters_list, block_name="Parameters")
+                    parameters = parse_typed_list(context, parameters_list,
+                                                  only_variables=True)
+                    precondition_tag_opt = next(iterator)
+                else:
+                    parameters = []
+                    precondition_tag_opt = parameters_tag_opt
+            with context.layer("Parsing precondition"):
+                if precondition_tag_opt == ":precondition":
+                    precondition_list = next(iterator)
+                    context.check_list(precondition_list, block_name="Precondition")
+                    if not precondition_list:
+                        # Note that :precondition () is allowed in PDDL.
+                        precondition = pddl.Conjunction([])
+                    else:
+                        precondition = parse_condition(
+                            context, precondition_list, type_dict, predicate_dict)
+                    effect_tag = next(iterator)
+                else:
+                    precondition = pddl.Conjunction([])
+                    effect_tag = precondition_tag_opt
+            with context.layer("Parsing effect"):
+                context.check_string(effect_tag, value=":effect", item_name="Effect tag",
+                                     finalmsg=f"Expecting '{syntax}'.")
+                effect_list = next(iterator)
+                context.check_list(effect_list, block_name="Effect")
+                eff = []
+                if effect_list:
+                    try:
+                        cost = parse_effects(
+                            context, effect_list, eff, type_dict, predicate_dict)
+                    except ValueError as e:
```

Comment:
Where does the ValueError come from? I think it should be caught more locally and changed to a ParserError in that context.